### PR TITLE
allow extra params in module set params

### PR DESCRIPTION
**Pull request by jmerkow**
Very simple argument pass-thru to the executor.
If you rename/remove layers.

### CODE REVIEW

**Comment by piiswrong** [comment]
allow_extra_params -> allow_extra.

File: python/mxnet/module/executor_group.py [hunk 1]
```diff
@@ -330,7 +330,7 @@ def reshape(self, data_shapes, label_shapes):
             self._default_execs = [i for i in self.execs]
         self.bind_exec(data_shapes, label_shapes, reshape=True)
 
-    def set_params(self, arg_params, aux_params):
+    def set_params(self, arg_params, aux_params, allow_extra_params=False):
```
Comment (piiswrong) [inline python/mxnet/module/executor_group.py]: forgot to rename this?

Maybe add a test case?

**Comment by jmerkow** [comment]
@piiswrong Done. I kept allow_extra_parms in `python/mxnet/module/executor_group.py` to match `executor.py`.  Happy to change that too, if you want them all to be `allow_params`

**Comment by piiswrong** [comment]
But you called it with allow_extra in module

**Comment by jmerkow** [comment]
I was trying to make executor and executor group consistent.  I will update.

**Comment by jmerkow** [comment]
fyi: https://github.com/dmlc/mxnet/blob/master/python/mxnet/model.py#L414

**Comment by jmerkow** [comment]
I can add a test case. Current init_params is only called with arg_params in `unit_test/test_module.py`
set_params in used in `gpu/test_operator_gpu.py`

Im thinking `unit_test/test_module.py` is best.

**Comment by piiswrong** [comment]
Thanks.
Do you have time to add a test case? Otherwise I can merge this in first.

**Comment by jmerkow** [comment]
It might take some time to get done.
I can swing back and do a quick and dirty one in a few weeks.

**Comment by jmerkow** [comment]
@piiswrong Is there a pattern you use to check failure modes? (like assertRaises in unit test).

Im drafting a test:

```
def test_set_params():
    # data iter
    mx.random.seed(11)
    data = mx.nd.array([[0.05, .10]]);
    label = mx.nd.array([[.01, 0.99]]);
    train_data = mx.io.NDArrayIter(data, label, batch_size=1)

    # symbols
    x = mx.symbol.Variable('data')
    x = mx.symbol.FullyConnected(name='fc_0', data=x, num_hidden=2)
    x = mx.symbol.Activation(name="act_0", data=x, act_type='sigmoid')
    x = mx.symbol.FullyConnected(name='fc_1', data=x, num_hidden=2)
    x = mx.symbol.Activation(name="act_1", data=x, act_type='sigmoid')
    x = mx.symbol.LinearRegressionOutput(data=x, name='softmax', grad_scale=2)


    # create module
    mod = mx.mod.Module(x, context=[mx.cpu()]);
    mod.bind(train_data.provide_data, label_shapes=train_data.provide_label,
                    for_training=True)
    arg_params = {'fc_2_weight'  : mx.nd.array([.60, .60])}
    
    # Should raise an error
    mod.set_params(arg_params=arg_params, aux_params={}, allow_missing=True, allow_extra=False)

    # Should not raise an error
    mod.set_params(arg_params=arg_params, aux_params={}, allow_missing=True, allow_extra=True)

    # TODO: Test allow missing (might as well)
    # TODO: Test both 'allow' == False
```

**Comment by piiswrong** [comment]
You can do 
```
try:
    fail()
except:
   return
raise xxx

**Comment by jmerkow** [comment]
what about something like:

```
def assertRaises(expected_exception, func, *args, **kwargs):
    
    try:
        func(*args, **kwargs)
    except expected_exception as e:
        pass
    else:
        assert False, "%s did not raise %s" % (func.__name__, expected_exception.__name__) # Did not raise proper exception
        

def test_set_params():    
    # data iter
    mx.random.seed(11)
    data = mx.nd.array([[0.05, .10]]);
    label = mx.nd.array([[.01, 0.99]]);
    train_data = mx.io.NDArrayIter(data, label, batch_size=1)

    # symbols
    x = mx.symbol.Variable('data')
    x = mx.symbol.FullyConnected(name='fc_0', data=x, num_hidden=2)
    x = mx.symbol.Activation(name="act_0", data=x, act_type='sigmoid')
    x = mx.symbol.FullyConnected(name='fc_1', data=x, num_hidden=2)
    x = mx.symbol.Activation(name="act_1", data=x, act_type='sigmoid')
    x = mx.symbol.LinearRegressionOutput(data=x, name='softmax', grad_scale=2)


    # create module
    mod = mx.mod.Module(x, context=[mx.cpu()]);
    mod.bind(train_data.provide_data, label_shapes=train_data.provide_label,
                    for_training=True)
    arg_params = {'fc_2_weight'  : mx.nd.array([.60, .60])}
    
    assertRaises(RuntimeError, mod.set_params, arg_params=arg_params, aux_params={}, allow_missing=True, allow_extra=True)
```

That way we can chain them in one function.  

P.S. is there any reason that unittest was avoided?  It's a very nice library.

**Comment by jmerkow** [comment]
@piiswrong I added a first draft of tests.  

FYI. if you set `allow_missing=False`, you do not need `allow_extra` at all, when there are extra args.  `allow_extra` is really only needed when there are missing+extra args.  
Practically, this makes sense, say you are fine tuning a new network from old params, where you have removed+added layers.

**Comment by piiswrong** [comment]
Thanks!

**Pull request merged**